crates/metadata/src/compatibility.rs

```diff
@@ -0,0 +1,161 @@
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct Requirements {
+    #[serde(rename = "ink")]
+    ink_requirements: Vec<VersionReq>,
+}
+
+/// Checks whether the contract's ink! version is compatible with the cargo-contract
+/// binary.
+pub fn check_contract_ink_compatibility(ink_version: &Version) -> Result<()> {
+    let compatibility_list = include_str!("../compatibility_list.json");
+    let compatibility: Compatibility = serde_json::from_str(compatibility_list)?;
+    let cargo_contract_version =
+        semver::Version::parse(VERSION).expect("Parsing version failed");
+    let ink_req = &compatibility
+        .cargo_contract_compatibility
+        .get(&cargo_contract_version)
+        .ok_or(anyhow!(
+            "Missing compatibility configuration for cargo-contract: {}",
+            cargo_contract_version
+        ))?
+        .ink_requirements;
+
+    // Check if the ink! version matches any of the requirement
+    if !ink_req.iter().any(|req| req.matches(ink_version)) {
+        // Get matching ink! versions
+        let ink_matches = ink_req
+            .iter()
+            .map(|req| format!("'{}'", req))
+            .collect::<Vec<_>>()
+            .join(", ");
+
+        let ink_matches_message = if !ink_matches.is_empty() {
+            format!("update the contract ink! to a version of {}", ink_matches)
+        } else {
+            String::default()
+        };
+
+        // Find best cargo-contract version
+        let cargo_contract_match_message = compatibility
+            .cargo_contract_compatibility
+            .iter()
+            .filter_map(|(ver, reqs)| {
+                if reqs
+                    .ink_requirements
+                    .iter()
+                    .any(|req| req.matches(ink_version))
+                {
+                    return Some(ver)
+                }
+                None
+            })
+            .max_by(|&a, &b| {
+                match (!a.pre.is_empty(), !b.pre.is_empty()) {
+                    (false, true) => std::cmp::Ordering::Greater,
+                    (true, false) => std::cmp::Ordering::Less,
+                    (_, _) => a.cmp(b),
+                }
+            })
+            .map(|ver| format!("update the cargo-contract to version '{}'", ver))
+            .unwrap_or_default();
+
+        let matches_message = [cargo_contract_match_message, ink_matches_message]
+            .into_iter()
+            .filter(|m| !m.is_empty())
+            .collect::<Vec<_>>()
+            .join(" or ");
+        if !matches_message.is_empty() {
+            return Err(anyhow!(
+                "The cargo-contract is not compatible with the contract's ink! version. Please {}",
+                 matches_message
+            ))
+        } else {
+            bail!(
+                "The cargo-contract is not compatible with the contract's ink! version, \
+                but a matching version could not be found"
+            )
+        }
```

Comment:
Then this code could be made clearer e.g. 

```rust
if let Some(cargo_contract_version) = best_cargo_contract_version {
   // build message
} else {
  bail!(
     "The cargo-contract is not compatible with the contract's ink! version, \
      but a matching version could not be found"
  )
}
```

Or even use `ok_or(bail!("The cargo-contract is not compatible ..."))?` on the end of the `best_cargo_contract_version` expression I suggest.

Reply:
done